driver-core/src/main/com/mongodb/internal/tracing/TracingManager.java

```diff
@@ -0,0 +1,241 @@
+     * @param namespace     The MongoDB namespace associated with the operation.
+     * @return The created span.
+     */
+    public Span addSpan(final String name, @Nullable final TraceContext parentContext, final MongoNamespace namespace) {
+        return tracer.nextSpan(name, parentContext, namespace);
+    }
+
+    /**
+     * Creates a new transaction span for the specified server session.
+     *
+     * @return The created transaction span.
+     */
+    public Span addTransactionSpan() {
+        Span span = tracer.nextSpan("transaction", null, null);
+        span.tagLowCardinality(SYSTEM.withValue("mongodb"));
+        return span;
+    }
+
+    /**
+     * Checks whether tracing is enabled.
+     *
+     * @return True if tracing is enabled, false otherwise.
+     */
+    public boolean isEnabled() {
+        return tracer.enabled();
+    }
+
+    /**
+     * Checks whether command payload tracing is enabled.
+     *
+     * @return True if command payload tracing is enabled, false otherwise.
+     */
+    public boolean isCommandPayloadEnabled() {
+        return enableCommandPayload;
+    }
+
+
+    /** Create a tracing span for the given command message.
+     * <p>
+     * The span is only created if tracing is enabled and the command is not security-sensitive.
+     * It attaches various tags to the span, such as database system, namespace, query summary, opcode,
+     * server address, port, server type, client and server connection IDs, and, if applicable,
+     * transaction number and session ID.
+     * If command payload tracing is enabled, the command document is also attached as a tag.
+     *
+     * @param message          the command message to trace
+     * @param operationContext the operation context containing tracing and session information
+     * @param commandDocumentSupplier a supplier that provides the command document when needed
+     * @param isSensitiveCommand a predicate that determines if a command is security-sensitive based on its name
+     * @param serverAddressSupplier a supplier that provides the server address when needed
+     * @param connectionIdSupplier a supplier that provides the connection ID when needed
+     * @return the created {@link Span}, or {@code null} if tracing is not enabled or the command is security-sensitive
+     */
+    @Nullable
+    public Span createTracingSpan(final CommandMessage message,
+            final OperationContext operationContext,
+            final Supplier<BsonDocument> commandDocumentSupplier,
+            final Predicate<String> isSensitiveCommand,
+            final Supplier<ServerAddress> serverAddressSupplier,
+            final Supplier<ConnectionId> connectionIdSupplier
+            ) {
+
+        BsonDocument command = commandDocumentSupplier.get();
+        String commandName = command.getFirstKey();
+        if (!isEnabled() || isSensitiveCommand.test(commandName)) {
+            return null;
+        }
```

Comment:
We currently perform [document deserialization from `BsonOutput`](https://github.com/mongodb/mongo-java-driver/blob/9a3696edeaf1cb4d36a1c501786852aa416e5999/driver-core/src/main/com/mongodb/internal/connection/InternalStreamConnection.java#L445) regardless of whether it’s enabled. Let’s move the `isEnabled` check before invoking the supplier.

```suggestion
       if (!isEnabled()) {
            return null;
        }
        BsonDocument command = commandDocumentSupplier.get();
        String commandName = command.getFirstKey();
        if (isSensitiveCommand.test(commandName)) {
            return null;
        }
```